content/blog/2019-11-13-gep.md/index.md

```diff
@@ -0,0 +1,458 @@
+allocation. 
+
+Additionally, we run our transformation as a function pass so
+it doesn't track interprocedural allocations. The main
+reason for this limitation is that, even if we knew
+the original allocation size for all callers, we would
+have to modify the function signature to communicate
+legal index values from the caller.
+This seemed both out of scope for our current project
+and a potentially questionable design decision. Should
+a compiler pass be modifying the signatures of
+potentially every function?
+
+Consequently, our pass is unable to improve the
+memory safety of this function:
+```C
+int foo(int* x, int y) {
+    return x[y];
+}
+```
+
+We had hoped to use LLVM's alias analysis or copy propagation tools
+to increase the precision of allocation tracking. However, we couldn't get
+these to work; they were difficult to integrate and didn't seem to track
+pointer value propagation as we expected them to. LLVM's relatively new
+[MemorySSA](https://llvm.org/docs/MemorySSA.html) analysis seemed very
+promising, since their example code finds domination relationships between
+memory uses and definitions. This would allow us to, at least for some cases,
+track allocation size information transitively through pointer reads and writes.
+However, the implementation is less precise than the documentation lets on
+and does not find accurate enough relationships to identify the root allocation
+for any pointers in practice.
+
+### Bitcasting
+Additionally, `bitcast` instructions complicate this process
+even more, since they cause the "sizes" of memory allocations to
+be interpreted differently.
+
+```C
+%1 = alloca i32, i64 10
+%2 = bitcast i32* %1 to i8*
+```
+
+Since 4 `i8` values fit into one `i32`, the allocation of `%2` represents a totally
+different number of elements than `%1` even though they represent the result of the
+same allocation operation.
+In the above code, a GEP that uses `%1` can safely index into elements 0 to 9.
+However, a GEP that uses `%2` as the base can safely index into elements 0 to 31.
+For any bitcast instruction that casts an allocation of known size, we convert
+and track the size of the new value, using integer multiplication and division
+to soundly approximate the maximum safe index. For typical bitcasts (e.g. `char` to `int`)
+this will not lose precision; however LLVM does have arbitrary precision integers,
+which could cause this estimate of allocation size to be an underestimate.
+
+
+## Soundness and Completeness
+
+Often, when trying to ensure a safety
+property you'd like to show that your results are either *sound* or *complete*.
+In our case, soundness would imply that any LLVM program which uses no
+"type unsafe" features and is compiled with our pass 
+executes *no GEP instructions which would generate out-of-bounds pointers*.
+Completeness, on the other hand, would imply that we can compile
+all programs and allow them to execute.
```

Comment:
I would actually frame "completeness" in this context _slightly_ differently: Whenever one of our early exits occurred, the program was _actually_ about to do an illegal memory access. In other words, while the tool can "miss" bugs, when it signals a bug, one actually exists. I think this means the same thing but is a little more useful of a definition.